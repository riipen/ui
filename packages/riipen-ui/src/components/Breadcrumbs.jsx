import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import Typography from "./Typography";

class Breadcrumbs extends React.Component {
  static propTypes = {
    /**
     * The breadcrumb children.
     */
    children: PropTypes.node.isRequired,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     * By default, it maps the variant to a good default headline component.
     */
    component: PropTypes.elementType,

    /**
     * Custom separator node.
     */
    separator: PropTypes.node
  };

  static defaultProps = {
    component: "nav",
    separator: "/"
  };

  static contextType = ThemeContext;

  render() {
    const { children, component, separator, ...other } = this.props;

    const theme = this.context;

    return (
      <React.Fragment>
        <Typography component={component} {...other}>
          <ol>
            {children.map((child, index) => {
              if (index < children.length - 1) {
                return [
                  <li key={`child-${index}`}>{child}</li>,
                  <li key={`separator-${index}`}>{separator}</li>
                ];
              }

              return <li key={`child-${index}`}>{child}</li>;
            })}
          </ol>
        </Typography>
        <style jsx>{`
          ol {
            align-items: center;
            display: flex;
            flex-wrap: wrap;
            margin: 0;
            padding: 0;
          }

          li {
            list-style: none;
            margin-right: ${theme.spacing(2)}px;
          }
          li::last-child {
            margin-right: 0;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(Breadcrumbs);
