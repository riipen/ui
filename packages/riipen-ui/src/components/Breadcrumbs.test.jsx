import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Breadcrumbs from "./Breadcrumbs.jsx";

describe("<Breadcrumbs>", () => {
  it("renders without errors", () => {
    let error;
    const child = <h1>hello hello</h1>;
    try {
      mount(<Breadcrumbs children={child} />);
    } catch (e) {
      error = e;
    }
    expect(error).toEqual(undefined);
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const child = <h1>hello hello</h1>;
      const defaultProps = new Breadcrumbs({ children: child }).type
        .defaultProps;

      const wrapper = mount(<Breadcrumbs children={child} />);

      const component = wrapper.find("Breadcrumbs");
      expect(component.props().component).toEqual(defaultProps.component);
      expect(component.props().separator).toEqual(defaultProps.separator);

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe("children prop", () => {
    it("renders children nodes as typography", () => {
      const child1 = <h1>hello hello</h1>;
      const child2 = <h1>hello 2</h1>;

      const wrapper = mount(
        <Breadcrumbs>
          {child1}
          {child2}
        </Breadcrumbs>
      );

      const list = wrapper
        .find("Breadcrumbs")
        .find("Typography")
        .find("ol")
        .find("li");
      expect(list.at(0).contains(child1)).toEqual(true);
      expect(list.at(2).contains(child2)).toEqual(true);
    });

    it("assigns children nodes unique keys", () => {
      const child1 = <h1>hello hello</h1>;
      const child2 = <h1>hello 2</h1>;

      const wrapper = mount(
        <Breadcrumbs>
          {child1} {child2}
        </Breadcrumbs>
      );

      const list = wrapper
        .find("Breadcrumbs")
        .find("Typography")
        .find("ol")
        .find("li");

      expect(list.at(0).key()).not.toEqual(list.at(2).key());
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the ol element", () => {
      const classes = ["classOne", "classTwo"];
      const child = <h1>hello hello</h1>;

      const wrapper = mount(<Breadcrumbs classes={classes} children={child} />);

      expect(
        wrapper
          .find("Breadcrumbs")
          .childAt(0)
          .find("ol")
          .hasClass(classes[0])
      ).toEqual(true);
      expect(
        wrapper
          .find("Breadcrumbs")
          .childAt(0)
          .find("ol")
          .hasClass(classes[1])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-breadcrumbs"];
      const child = <h1>hello hello</h1>;

      const wrapper = mount(<Breadcrumbs children={child} />);

      expect(
        wrapper
          .find("Breadcrumbs")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const child = <h1>hello hello</h1>;
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Breadcrumbs classes={classes} children={child} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("component prop", () => {
    it("passes given value into the component prop of Typography", () => {
      const component = "h1";

      const child = <h1>hello hello</h1>;

      const wrapper = mount(
        <Breadcrumbs component={component} children={child} />
      );

      expect(
        wrapper
          .find("Breadcrumbs")
          .find("Typography")
          .props().component
      ).toEqual(component);
    });
  });

  describe("separator prop", () => {
    it("separates children nodes with default separator", () => {
      const child1 = <h1>hello hello</h1>;
      const child2 = <h1>hello 2</h1>;
      const defaultSeparator = new Breadcrumbs({ children: [child1, child2] })
        .type.defaultProps.separator;

      const wrapper = mount(
        <Breadcrumbs>
          {child1}
          {child2}
        </Breadcrumbs>
      );

      const list = wrapper
        .find("Breadcrumbs")
        .find("Typography")
        .find("ol")
        .find("li");
      expect(list.at(0).contains(child1)).toEqual(true);
      expect(list.at(1).text()).toEqual(defaultSeparator);
      expect(list.at(2).contains(child2)).toEqual(true);
    });

    it("separates children nodes with custom separator", () => {
      const child1 = <h1>hello hello</h1>;
      const child2 = <h1>hello 2</h1>;
      const separator = "mmm";

      const wrapper = mount(
        <Breadcrumbs separator={separator}>
          {child1}
          {child2}
        </Breadcrumbs>
      );

      const list = wrapper
        .find("Breadcrumbs")
        .find("Typography")
        .find("ol")
        .find("li");
      expect(list.at(0).contains(child1)).toEqual(true);
      expect(list.at(1).text()).toEqual(separator);
      expect(list.at(2).contains(child2)).toEqual(true);
    });
  });
});
