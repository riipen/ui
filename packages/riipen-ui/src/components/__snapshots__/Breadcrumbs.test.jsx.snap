// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs> default props sets correct default props 1`] = `
<withClasses(Breadcrumbs)>
  <Breadcrumbs
    classes={
      Array [
        "riipen",
        "riipen-breadcrumbs",
      ]
    }
    component="nav"
    separator="/"
  >
    <withClasses(Typography)
      component="nav"
    >
      <Typography
        classes={
          Array [
            "riipen",
            "riipen-typography",
          ]
        }
        color="inherit"
        component="nav"
        display="initial"
        gutter={false}
        textAlign="inherit"
        variant="body1"
      >
        <nav
          className="jsx-4149659077 root color-inherit initial body1 align-inherit riipen riipen-typography"
        >
          <ol
            className="jsx-2183045356 riipen riipen-breadcrumbs"
          >
            <li
              className="jsx-2183045356"
              key="child-0/.0"
            >
              <h1>
                hello hello
              </h1>
            </li>
          </ol>
        </nav>
        <JSXStyle
          dynamic={
            Array [
              "#424242",
              "Roboto, Helvetica, Arial, sans-serif",
              400,
              "14px",
              "20px",
              "0.00938em",
              "#424242",
              "Roboto, Helvetica, Arial, sans-serif",
              400,
              "12px",
              "17px",
              "0.01071em",
              "#424242",
              "Roboto, Helvetica, Arial, sans-serif",
              400,
              "10px",
              "14px",
              "0.01071em",
              "#373737",
              "Roboto, Helvetica, Arial, sans-serif",
              500,
              "40px",
              "56px",
              "-0.01562em",
              "#373737",
              "Roboto, Helvetica, Arial, sans-serif",
              500,
              "30px",
              "42px",
              "-0.00833em",
              "#373737",
              "Roboto, Helvetica, Arial, sans-serif",
              500,
              "22px",
              "32px",
              "0em",
              "#373737",
              "Roboto, Helvetica, Arial, sans-serif",
              400,
              "20px",
              "30px",
              "-0.00735em",
              "#373737",
              "Roboto, Helvetica, Arial, sans-serif",
              400,
              "16px",
              "23px",
              "0rem",
              "#000",
              "#747474",
              "#424242",
              "#373737",
              "#1f71ae",
              "#f58632",
              "#3caabb",
              "#4cc88d",
              "#e95353",
              "#fff",
              700,
              300,
              500,
              400,
            ]
          }
          id="489809676"
        />
      </Typography>
    </withClasses(Typography)>
    <JSXStyle
      dynamic={
        Array [
          10,
        ]
      }
      id="2237344651"
    />
  </Breadcrumbs>
</withClasses(Breadcrumbs)>
`;
